packages/belos/tpetra/src/solvers/Belos_Tpetra_GmresSstep.hpp

```diff
@@ -333,8 +369,26 @@ class GmresSstep : public Gmres<SC, MV, OP>  {
         }
 
         // Orthogonalization
-        this->projectBelosOrthoManager (iter, step, Q, G);
-        const int rank = normalizeCholQR (iter, step, Q, G);
+        {
+          Teuchos::TimeMonitor LocalTimer (*bortTimer);
+          this->projectBelosOrthoManager (iter, step, Q, G);
+        }
+        int rank = 0;
+        {
+          Teuchos::TimeMonitor LocalTimer (*tsqrTimer);
+          rank = normalizeCholQR (iter, step, Q, G);
+          if (useCholQR2_) {
+            rank = normalizeCholQR (iter, step, Q, G2);
+            // merge R 
+            dense_matrix_type Rfix (Teuchos::View, G2, step+1, step+1, iter, 0);
+            dense_matrix_type Rold (Teuchos::View, G,  step+1, step+1, iter, 0);
+            blas.TRMM (Teuchos::LEFT_SIDE, Teuchos::UPPER_TRI,
```

Comment:
It looks like the issue is that you want to use `Belos::(Mat)OrthoManager`, but that interface always takes small dense matrices as host data in a `Teuchos::SerialDenseMatrix`.  This is why you're calling the BLAS on host here.  Is that right?

Reply:
I was doing it just for convenience (e.g., I convert R to H, and solve the small least-square problem on the host), but I think you are right. We have the option to use Belos OrthoManager, which keeps the small matrices on host. I'll look at if it makes more sense to keep them on device for the performance..